datanode/snapshot/service_create_snapshot.go

```diff
@@ -0,0 +1,257 @@
+	}
+
+	_, err = copyDataTx.Exec(ctx, "SET TRANSACTION ISOLATION LEVEL SERIALIZABLE")
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to set transaction isolation level to serilizable:%w", err)
+	}
+
+	writeLockFileName := filepath.Join(b.snapshotsPath, WriteLockFileName(chainID, toHeight))
+	_, err = os.Create(writeLockFileName)
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to create write lock file:%w", err)
+	}
+
+	go func() {
+		b.log.Infof("copying all table data....")
+		allCopySQL := append(currentSnapshot.GetCopySQL(dbMetaData, b.config.DatabaseSnapshotsPath),
+			historySnapshot.GetCopySQL(dbMetaData, b.config.DatabaseSnapshotsPath)...)
+		rowsCopied, err := copyTableData(ctx, copyDataTx, allCopySQL)
+		if err != nil {
+			b.logError(fmt.Errorf("failed to copy table data:%w", err))
+			err = copyDataTx.Rollback(ctx)
+			if err != nil {
+				b.logError(fmt.Errorf("failed to rollback snapshot transaction:%w", err))
+			}
+
+			return
+		}
+
+		err = copyDataTx.Commit(ctx)
+		if err != nil {
+			b.logError(fmt.Errorf("failed to commit snapshot transaction:%w", err))
+			return
+		}
+
+		b.log.Infof("compressing current state snapshot data")
+		compressedCurrentStateByteCount, err := tarAndCompressSnapshot(b.snapshotsPath, currentSnapshot, dbMetaData.DatabaseVersion)
+		if err != nil {
+			b.logError(fmt.Errorf("failed to compress snapshot:%w", err))
+			return
+		}
+		b.log.Infof("compressed current state snapshot data size: %dB", compressedCurrentStateByteCount)
+
+		b.log.Infof("compressing history data")
+		compressedHistoryByteCount, err := tarAndCompressSnapshot(b.snapshotsPath, historySnapshot, dbMetaData.DatabaseVersion)
+		if err != nil {
+			b.logError(fmt.Errorf("failed to compress history snapshot:%w", err))
+			return
+		}
+		b.log.Infof("compressed history snapshot data size: %dB", compressedHistoryByteCount)
+
+		if err := os.RemoveAll(currentDataDir); err != nil {
+			b.logError(fmt.Errorf("failed to remove uncompressed current data directory  %s :%w", currentDataDir, err))
+			return
+		}
+
+		if err := os.RemoveAll(historyDataDir); err != nil {
+			b.logError(fmt.Errorf("failed to remove uncompressed current data directory  %s :%w", historyDataDir, err))
+			return
+		}
+
+		err = os.Remove(writeLockFileName)
+		if err != nil {
+			b.logError(fmt.Errorf("failed to remove write lock file:%w", err))
+			return
+		}
+
+		snapshotHash, err := GetSnapshotMd5Hash(b.log, meta)
+		if err != nil {
+			b.logError(fmt.Errorf("failed to get md5 hash of snapshot:%w", err))
+		}
+
+		metrics.SetLastSnapshotRowcount(float64(rowsCopied))
+		metrics.SetLastSnapshotCurrentStateBytes(float64(compressedCurrentStateByteCount))
+		metrics.SetLastSnapshotHistoryBytes(float64(compressedHistoryByteCount))
+		metrics.SetLastSnapshotSeconds(time.Since(start).Seconds())
+
+		b.log.Infof("finished creating snapshot for chain:%s, height:%d, time taken:%s, rows copied:%d, total compressed data size:%d, md5 hash:%s", chainID, toHeight,
```

Comment:
could we use structured logging in general? we do everywhere, makes things easier to parse usually.